src/nbl/video/IGPUCommandBuffer.cpp

```diff
@@ -0,0 +1,951 @@
+    uint32_t imageMemoryBarrierCount, const SImageMemoryBarrier* pImageMemoryBarriers)
+{
+    if ((memoryBarrierCount == 0u) && (bufferMemoryBarrierCount == 0u) && (imageMemoryBarrierCount == 0u))
+        return false;
+
+    constexpr auto MaxBarrierResourceCount = (1 << 12) / sizeof(void*);
+    assert(bufferMemoryBarrierCount + imageMemoryBarrierCount <= MaxBarrierResourceCount);
+
+    core::smart_refctd_ptr<const IGPUBuffer> bufferResources[MaxBarrierResourceCount];
+    for (auto i = 0; i < bufferMemoryBarrierCount; ++i)
+        bufferResources[i] = pBufferMemoryBarriers[i].buffer;
+
+    core::smart_refctd_ptr<const IGPUImage> imageResources[MaxBarrierResourceCount];
+    for (auto i = 0; i < imageMemoryBarrierCount; ++i)
+        imageResources[i] = pImageMemoryBarriers[i].image;
+
+    if (!m_cmdpool->emplace<IGPUCommandPool::CPipelineBarrierCmd>(m_segmentListHeadItr, m_segmentListTail, bufferMemoryBarrierCount, bufferResources, imageMemoryBarrierCount, imageResources))
+        return false;
+
+    pipelineBarrier_impl(srcStageMask, dstStageMask, dependencyFlags, memoryBarrierCount, pMemoryBarriers, bufferMemoryBarrierCount, pBufferMemoryBarriers, imageMemoryBarrierCount, pImageMemoryBarriers);
+
+    return true;
+}
+
+bool IGPUCommandBuffer::bindDescriptorSets(asset::E_PIPELINE_BIND_POINT pipelineBindPoint, const pipeline_layout_t* layout, uint32_t firstSet, const uint32_t descriptorSetCount,
+    const descriptor_set_t* const* const pDescriptorSets, const uint32_t dynamicOffsetCount, const uint32_t* dynamicOffsets)
+{
+    if (!this->isCompatibleDevicewise(layout))
+        return false;
+
+    if (layout->getAPIType() != getAPIType())
+        return false;
+
+    for (uint32_t i = 0u; i < descriptorSetCount; ++i)
+    {
+        if (pDescriptorSets[i])
+        {
+            if (!this->isCompatibleDevicewise(pDescriptorSets[i]))
+            {
+                m_logger.log("IGPUCommandBuffer::bindDescriptorSets failed, pDescriptorSets[%d] was not created by the same ILogicalDevice as the commandbuffer!", system::ILogger::ELL_ERROR, i);
+                return false;
+            }
+            if (!pDescriptorSets[i]->getLayout()->isIdenticallyDefined(layout->getDescriptorSetLayout(firstSet + i)))
+            {
+                m_logger.log("IGPUCommandBuffer::bindDescriptorSets failed, pDescriptorSets[%d] not identically defined as layout's %dth descriptor layout!", system::ILogger::ELL_ERROR, i, firstSet + i);
+                return false;
+            }
+        }
+    }
+
+    core::smart_refctd_ptr<const video::IGPUDescriptorSet> descriptorSets_refctd[IGPUPipelineLayout::DESCRIPTOR_SET_COUNT] = { nullptr };
+    for (auto i = 0; i < descriptorSetCount; ++i)
+    {
+        assert(i < IGPUPipelineLayout::DESCRIPTOR_SET_COUNT);
+        descriptorSets_refctd[i] = core::smart_refctd_ptr<const video::IGPUDescriptorSet>(pDescriptorSets[i]);
+    }
```

Comment:
why not fold all this into the `CBindDescriptorSetsCmd` constructor?

Reply:
Sure.

Reply:
Resolved.